wasm-node/javascript/src/internals/client.ts

```diff
@@ -552,27 +552,45 @@ export function start(options: ClientOptions, wasmModule: SmoldotBytecode | Prom
+                jsonRpcResponses: {
+                    next: async () => {
+                        while (true) {
+                            if (!state.chains.has(chainId))
+                                return { done: true, value: undefined };
+                            if (options.disableJsonRpc)
+                                throw new JsonRpcDisabledError();
+                            if (state.instance.status === "destroyed")
+                                throw state.instance.error;
+                            if (state.instance.status !== "ready")
+                                throw new Error(); // Internal error. Never supposed to happen.
+    
+                            // Try to pop a message from the queue.
+                            const message = state.instance.instance.peekJsonRpcResponse(chainId);
+                            if (message)
+                                return { done: false, value: message };
+    
+                            // If no message is available, wait for one to be.
+                            await new Promise<void>((resolve) => {
+                                state.chains.get(chainId)!.jsonRpcResponsesPromises.push(resolve)
+                            });
+                        }
+                    },
+                    return: async () => {
+                        newChain.remove();
+                        return { done: true, value: undefined };
+                    },
+                    [Symbol.asyncIterator]() {
+                        return this
+                    }
+                },
                 nextJsonRpcResponse: async () => {
-                    while (true) {
-                        if (!state.chains.has(chainId))
-                            throw new AlreadyDestroyedError();
-                        if (options.disableJsonRpc)
-                            return Promise.reject(new JsonRpcDisabledError());
-                        if (state.instance.status === "destroyed")
-                            throw state.instance.error;
-                        if (state.instance.status !== "ready")
-                            throw new Error(); // Internal error. Never supposed to happen.
-
-                        // Try to pop a message from the queue.
-                        const message = state.instance.instance.peekJsonRpcResponse(chainId);
-                        if (message)
-                            return message;
+                    const result = await newChain.jsonRpcResponses.next();
 
-                        // If no message is available, wait for one to be.
-                        await new Promise<void>((resolve) => {
-                            state.chains.get(chainId)!.jsonRpcResponsesPromises.push(resolve)
-                        });
+                    if (result.done) {
+                        throw new AlreadyDestroyedError();
                     }
+
+                    return result.value;
```

Comment:
I would actually much prefer if this function was left as it is, and `jsonRpcResponses` was implemented on top of it. This function is much more simple than `jsonRpcResponses` from an implementation point of view.